include/nbl/asset/IDescriptorSetLayout.h

```diff
@@ -78,15 +81,68 @@ enum E_DESCRIPTOR_TYPE : uint32_t
 template<typename SamplerType>
 class NBL_API IDescriptorSetLayout : public virtual core::IReferenceCounted
 {
+	struct SBindingRedirect
+	{
+		static constexpr inline uint32_t Invalid = ~0u;
+
+		SBindingRedirect() : bindings(nullptr), offsets(nullptr), count(0ull) {}
+
+		SBindingRedirect(const size_t _count) : count(_count)
+		{
+			bindings = std::make_unique<uint32_t[]>(count << 1);
+			offsets = bindings.get() + count;
+		}
+
+		// For a given binding number `binding` in `bindings`, return its offset in `offsets`.
+		inline uint32_t operator[](const uint32_t binding) const
+		{
+			const uint32_t index = searchForBinding(binding);
+			if (index == Invalid)
+				return Invalid;
+
+			return offsets[index];
+		}
```

Comment:
I'd opt for a bit of type safety and do this
```cpp
struct binding_t
{
   uint32_t value;
};
struct storage_offset_t
{
   uint32_t value;
};
struct SBindingRedirect
{
```

then you could use `binding_t` and `storage_offset_t` instead of having `uint32_t` flying all over the place.
You could even give these struct implicit constructors from `uint32_t` to save yourself typing some `{`,`}` characters.

Reply:
with type safety you could actually have
```cpp
// find offset into storage based on descriptor binding
storage_offset_t operator[](const binding_t _binding) const;
// find descriptor binding based on storage offset
binding_t operator[](const storage_offset_t _offset) const;
```

Reply:
and `searchForBinding` could turn into `binding_t* operator[](const binding_t _binding) const;`

Reply:
Yes, `searchForBinding` was used extensively in `COpenGLDescriptorSet`. Now with that gone, I think I can just inline it.

Reply:
nice, but I think a function to find the pointer/iterator to the sorted binding point list would be useful

Reply:
Yeah useful, even currently. Resolved.